### PR TITLE
Make Grafana product performance dashboard support SSL

### LINKED ISSUE #973: Grafana data source not getting created when SSL is enabled
**Description:**

Grafana data source  is not getting created when SSL is enabled. So the dashboard cannot display the gathered perfomance metrics.

### PR DESCRIPTION
**Pull request by tharindu1992**
**Purpose**

Currently Grafana data source is not getting created when SSL is enabled. So the dashboard cannot display the gathered perfomance metrics. These changes will fix this issue
Resolves #973 

**Goals**

Create Grafana data sources successfully 

**Approach**

Bypassed host verification for Grafana URL.

**Security checks**
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java [hunk 1]
```diff
@@ -31,17 +28,28 @@
...
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLHandshakeException;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509TrustManager;
+
+import static org.apache.http.protocol.HTTP.USER_AGENT;
 
 /**
  * This class create a database in influxDB and a Data Source in
```
Comment (pasindujw) [inline common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: Incomplete comment

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java [hunk 2]
```diff
@@ -99,50 +108,68 @@ private String getDataSource(String name) {
...
+            con.setRequestMethod("POST");
+            con.setRequestProperty("User-Agent", USER_AGENT);
+            con.setRequestProperty("Content-Type", ContentType.APPLICATION_JSON.toString());
+            con.setRequestProperty("Accept", ContentType.APPLICATION_JSON.toString());
+            con.setRequestProperty("Authorization", apikey);
+            SSLSocketFactory sslSocketFactory = createSslSocketFactory();
+            con.setSSLSocketFactory(sslSocketFactory);
+            String urlParameters = getDataSource(testplanID);
+            con.setDoOutput(true);
+            //logger.info("garafana COnnection: " + con.toString());
```
Comment (pasindujw) [inline common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: Commented line of code

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java [hunk 3]
```diff
@@ -99,50 +108,68 @@ private String getDataSource(String name) {
...
-    }
+    public void addGrafanaDataSource() {
 
-    /**
-     * This command will handle the posting http request to REST  api
-     * @param jsonData json string of the data source
-     * @param httpPost post request with authentication
-     */
-    private void executeReq(String jsonData, HttpPost httpPost) {
+        DataOutputStream wr = null;
```
Comment (pasindujw) [inline common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: Shall we name the variable as it reflects the variable type?

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java [hunk 4]
```diff
@@ -99,50 +108,68 @@ private String getDataSource(String name) {
...
-     * @param httpPost post request with authentication
-     * @throws UnsupportedEncodingException thown when an error with api key
-     * @throws IOException thrown when error in connection
-     */
-    private void executeHttpRequest(String jsonData,  HttpPost httpPost)  throws UnsupportedEncodingException,
-            IOException {
-        httpPost.setEntity(new StringEntity(jsonData));
-        HttpClient client = HttpClientBuilder.create().build();
-        client.execute(httpPost);
+    private static SSLSocketFactory createSslSocketFactory() throws Exception {
```
Comment (pasindujw) [inline common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: Missing method comments

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java [hunk 5]
```diff
@@ -60,7 +68,7 @@ public DashboardSetup(String testplanID) {
     public void initDashboard() {
         // create influxDB database according to tp_id
         try {
-            InfluxDB influxDB = InfluxDBFactory.connect(restUrl + ":8086", username, password);
+            InfluxDB influxDB = InfluxDBFactory.connect("http://" + restUrl + ":8086", username, password);
```
Comment (azinneera) [inline common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: `"http://"`  string is used in several places. So better we declare it as a Constant.

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java [hunk 6]
```diff
@@ -99,50 +108,68 @@ private String getDataSource(String name) {
...
+            wr.flush();
+            int responseCode = con.getResponseCode();
+            if (responseCode == 200) {
+                logger.info("grafana Data Source created for testplan " + testplanID);
+            } else {
+                logger.error(StringUtil.concatStrings("failed to create grafana Data testplan ", testplanID,
+                        " Response Code ", responseCode));
+            }
+        } catch (SSLHandshakeException e) {
+            logger.error("Please set a valid truststore key file" + e.toString());
```
Comment (pasindujw) [inline common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: Is it possible to say exactly that having a invalid truststore key file is the reason for `SSLHandshakeException` ?

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java [hunk 7]
```diff
@@ -99,50 +108,68 @@ private String getDataSource(String name) {
...
-            logger.error(StringUtil.concatStrings("ioException occured while sending http request : ", e));
+            String url = "https://" + ConfigurationContext.getProperty(ConfigurationContext.ConfigurationProperties
+                    .GRAFANA_DATASOURCE) + ":3000/api/datasources/";
+
+            HostnameVerifier allHostsValid = (hostname, session) -> true;
+            HttpsURLConnection.setDefaultHostnameVerifier(allHostsValid);
+
+            URL obj = new URL(url);
+            HttpsURLConnection con = (HttpsURLConnection) obj.openConnection();
+            //add reuqest header
```
Comment (pasindujw) [inline common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: Typo  in the comment

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java (hunk 6 above)
Comment (azinneera) [inline common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: Converting the exception to String is not necessary. logger.error accepts the Throwable. Second argument can be changed to just `e`.

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java (hunk 4 above)
Comment (tharindu1992) [reply common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: Comments added

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java [hunk 8]
```diff
@@ -18,11 +18,8 @@
 
 package org.wso2.testgrid.common;
 
-import org.apache.http.client.HttpClient;
-import org.apache.http.client.methods.HttpPost;
+
```
Comment (yasassri) [inline common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: Unnecessary blank line.

File: common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java [hunk 9]
```diff
@@ -99,50 +108,70 @@ private String getDataSource(String name) {
...
+            con.setRequestProperty("Authorization", apikey);
+            SSLSocketFactory sslSocketFactory = createSslSocketFactory();
+            con.setSSLSocketFactory(sslSocketFactory);
+            String urlParameters = getDataSource(testplanID);
+            con.setDoOutput(true);
+            dataOutputStream = new DataOutputStream(con.getOutputStream());
+            dataOutputStream.writeBytes(urlParameters);
+            dataOutputStream.flush();
+            int responseCode = con.getResponseCode();
+            if (responseCode == 200) {
```
Comment (pasindujw) [inline common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: Is it possible to refer the 200 value from a relevant reference class?
Ex:  `javax.ws.rs.core.Response.Status.Ok`

Comment (tharindu1992) [reply common/src/main/java/org/wso2/testgrid/common/DashboardSetup.java]: will need to add new dependancies

**Pull request merged**